Using a topline to provide flexibility of topline to match

M22an opened the pull request:
There are some cases we have where we want to ensure that the overall sum remains the same not on the sum of input, but based on another topline. 

I have made the code changes, by default the existing logic of retaining the sum of inputs is maintained, **topline** is added as an optional parameter.

Commit:
Using a topline to provide flexibility of topline to match instead of the sum of the original records

cgdeboer commented:
@M22an thanks for the contribution ! 

This is an interesting use case, at some point, I'd like to learn a bit more about the domain problem you are trying to solve with it, particularly when the `topline` differs from the sum of the original by a large amount.

I'm leaving some code comments on here for you now.

cgdeboer commented on iteround/core.py:
```diff
@@ -6,7 +6,7 @@
 
 
-def saferound(iterable, places, strategy=DIFFERENCE, rounder=round):
+def saferound(iterable, places, topline=-1, strategy=DIFFERENCE, rounder=round):
```
`topline=None` would be more idiomatic here.

cgdeboer commented on iteround/core.py:
```diff
@@ -65,7 +65,7 @@ def saferound(iterable, places, strategy=DIFFERENCE, rounder=round):
     # calculate original sum, rounded,  then rounded local sum.
     local = [Number(i, value) for i, value in enumerate(values)]
-    orig_sum = _sumnum(local, places, rounder)
+    orig_sum = _sumnum(local, places, rounder) if topline == -1 else topline
```
Style semantics, but if you change `topline`'s default value to `None`, then we can use the following pythonic expression.

`orig_sum = topline or _sumnum(local, places, rounder)`

cgdeboer reviewed (changes requested):
As mentioned above, it's an interesting addition. Please address the following:

- [x] address code comments that I've made below.
- [x] write a new test that uses `topline` 
- [x] add something to the README that discusses `topline` as an option.

M22an commented:
> @M22an thanks for the contribution !
> 
> This is an interesting use case, at some point, I'd like to learn a bit more about the domain problem you are trying to solve with it, particularly when the `topline` differs from the sum of the original by a large amount.
> 
> I'm leaving some code comments on here for you now.

First of all thanks for the quick turn-around.

This is actually a specific use case where we make changes to some values (changes like making zero as 1, due to a business rule). But we however don't want to affect the original sum, which we have before applying these business rules. 

You are right, in that, if the difference is too large that could cause problems as well, but we are trying to tackle it outside of iteround, as those a business specific rules. 

I will make the changes that you have highlighted and push.

Commit:
Changes based on comments, tests
Adding a rounder(topline) to handle topline being in different precision than required output

Commit:
Adding places to the rounder method call for topline

M22an commented:
@cgdeboer , I have pushed again with the changes you suggested. In addition to that I have made one change where I have used `rounder(topline, places)` to avoid rounding issues when comparing against overall adjusted sum.

cgdeboer commented:
@M22an looks good. Last 🙏 request. Can you squash these commits down to one with a nice message like "adds topline functionality. " also, add yourself to the `AUTHORS.rst` file 💯 

After those are done, I'll merge and work on pushing this out to pypi as version 1.0.3

Commit:
Changing the Authors.rst file

M22an commented:
@cgdeboer , I am a newbie to GitHub, so not sure how to squash my commits separately, but in other repos like Gitlab, I could always squash commits during merge. Can you please check if that's possible in GitHub as well ?

If not, I think I need to create a new fork and create new commits there.

cgdeboer commented:
@M22an no worries ! I'll handle it post merge... thanks for the contribution.

***

"Squashing" (and rebasing more broadly) can be an important skill when working in teams, and contributing to open source. Learning how to do it with `git` on the command line can be a valuable skill. Here's a tutorial for next time from the fine folks at git tower: https://www.git-tower.com/learn/git/faq/git-squash/

cgdeboer commented:
@M22an this is published on pypi as iteround `1.0.3`
